include/exadg/operators/generalized_laplace_operator.h

```diff
@@ -0,0 +1,849 @@
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+#ifndef INCLUDE_EXADG_OPERATORS_GENERALIZED_LAPLACE_OPERATOR_H_
+#define INCLUDE_EXADG_OPERATORS_GENERALIZED_LAPLACE_OPERATOR_H_
+
+#include <exadg/functions_and_boundary_conditions/evaluate_functions.h>
+#include <exadg/grid/grid_utilities.h>
+#include <exadg/operators/interior_penalty_parameter.h>
+#include <exadg/operators/operator_base.h>
+#include <exadg/operators/variable_coefficients.h>
+#include <exadg/poisson/user_interface/boundary_descriptor.h>
+#include <variant>
+
+namespace ExaDG
+{
+namespace GeneralizedLaplace
+{
+/**
+ * Free function for a generic expression of the multiplication with coefficients of different
+ * ranks.
+ */
+template<typename T, int coefficient_rank, int dim, typename Number>
+static inline DEAL_II_ALWAYS_INLINE //
+  T
+  coeff_mult(dealii::Tensor<coefficient_rank, dim, Number> const & coeff, T const & x)
+{
+  if constexpr(coefficient_rank == 4)
+    return dealii::double_contract<2, 0, 3, 1>(coeff, x);
+  else
+    return coeff * x;
+}
+
+enum class CoefficientModel
+{
+  ConstantFunction,
+  VariableFunction
+};
+
+/**
+ * Collection of data for the generalized Laplace kernel.
+ */
+template<int dim>
+struct KernelData
+{
+  //! Internal penalty factor for the calculation of `tau`
+  double IP_factor{1.0};
+
+  //!
+  bool calculate_and_store_cell_based_face_coefficients{false};
+
+  //!
+  CoefficientModel coefficient_model;
+
+  //!
+  std::shared_ptr<dealii::Function<dim>> coefficient_function;
+};
+
+/**
+ * Generalized Laplace kernel. This class is responsible for providing the core pieces of
+ * information required to evaluate the Laplace operator, i.e. the fluxes, numerical parameters
+ * (like the internal penalty parameter), and the coefficients.
+ *
+ * @tparam n_components Number of components of the solution (vector) field. Solution is scalar if
+ * this is equal to 1.
+ * @tparam coupling_coefficient Boolean switch to differentiate if the coefficient type couples
+ * the solution vector components or not.
+ *  - `false` always results in a scalar coefficient,
+ *  - `true` results in a 2-rank tensor coefficient if `n_components = 1` and in a 4-rank tensor
+ *  coefficient if `n_components > 1`.
+ */
+template<int dim, typename Number, int n_components = 1, bool coupling_coefficient = false>
```

Comment:
I find the name `coupling_coefficient` irritating. From the name, I would expect a coefficient, not a boolean. What about `coefficient_couples_components`?

Reply:
Yeah it's a bit irritating 🤣. `coefficient_couples_components` sounds better. Regarding your question from above https://github.com/exadg/exadg/pull/414#discussion_r1206431481 maybe we should even call it `bool coefficient_is_scalar` (with reversed true/false of course), because the case you described with a diagonal rank-2 tensor would have to fall into `coupling_coefficient=true` with the current implementation.